internal/lib/container_server.go

```diff
@@ -668,3 +668,52 @@ func (c *ContainerServer) StopContainerAndWait(ctx context.Context, ctr *oci.Con
 	}
 	return nil
 }
+
+func (c *ContainerServer) UpdateContainerLinuxResources(ctr *oci.Container, resources *rspec.LinuxResources) {
+	updatedSpec := ctr.Spec()
+	if updatedSpec.Linux == nil {
+		updatedSpec.Linux = &rspec.Linux{}
+	}
+
+	if updatedSpec.Linux.Resources == nil {
+		updatedSpec.Linux.Resources = &rspec.LinuxResources{}
+	}
+
+	if updatedSpec.Linux.Resources.CPU == nil {
+		updatedSpec.Linux.Resources.CPU = &rspec.LinuxCPU{}
+	}
+
+	if *resources.CPU.Shares != 0 {
+		updatedSpec.Linux.Resources.CPU.Shares = resources.CPU.Shares
+	}
+
+	if *resources.CPU.Quota != 0 {
+		updatedSpec.Linux.Resources.CPU.Quota = resources.CPU.Quota
+	}
+
+	if *resources.CPU.Period != 0 {
+		updatedSpec.Linux.Resources.CPU.Period = resources.CPU.Period
+	}
+
+	if resources.CPU.Cpus != "" {
+		updatedSpec.Linux.Resources.CPU.Cpus = resources.CPU.Cpus
+	}
+
+	if resources.CPU.Mems != "" {
+		updatedSpec.Linux.Resources.CPU.Mems = resources.CPU.Mems
+	}
+
+	if updatedSpec.Linux.Resources.Memory == nil {
+		updatedSpec.Linux.Resources.Memory = &rspec.LinuxMemory{}
+	}
+
+	if *resources.Memory.Limit != 0 {
+		updatedSpec.Linux.Resources.Memory.Limit = resources.Memory.Limit
+	}
+
+	updatedSpec.Linux.Resources.Memory.Swap = resources.Memory.Swap
+
+	ctr.SetSpec(&updatedSpec)
+
+	c.state.containers.Add(ctr.ID(), ctr)
```

Comment:
speaking about the general direction, this is how I would have done it myself (and AFAIK much more like containerd does). I don't know well enough cri-o internals to tell if we need locking here.
In general, my only concern is: should cri-o crash or restart (is this even a concern or is considered a "cannot happen" scenario?), do we need to care about restoring the correct value and updating the memory store? there is a recovery mechanism in place already?

Reply:
It already has the locking mechanism under the memory store - https://github.com/cri-o/cri-o/blob/d3bee57675a8b29a2774afcb3c662c605c8068d5/internal/oci/memory_store.go#L21

Regarding the restart or crash of the cri-o daemon, an interesting part that the updated spec survives it, I tried the scenario:
- start the pod
- check that the spec updated via `crictl inspect`
- restart the cri-o daemon
- check again the container spec, and it still has updated values

Reply:
yep, thanks for checking out the locking. Learned myself only after commenting (action on me: next time check before commenting :) ).

Reply:
One who was responsible to update the spec again was the CPU manager reconcile loop, it all time called `UpdateContainerResources`, so I unsure how feasible the memory store solution because we can rely on another component on the node to call  `UpdateContainerResources` after the crio crash or restart.

Reply:
I think it is runc who is keeping track of that information (that's how cri-o would have gotten it upon restoring), which is behavior I am fine with relying on

Reply:
CRI-O restores state back from config.json, its own state file and runtime state when it is restarted. If we think that the update reconcile logic is sufficient for this, then I am fine with this approach.